providers/provider_data.go

```diff
@@ -73,6 +71,99 @@ func (p *ProviderData) GetClientSecret() (clientSecret string, err error) {
 	return string(fileClientSecret), nil
 }
 
+// LoginURLParams returns the parameter values that should be passed to the IdP
+// login URL.  This is the default set of parameters configured for this provider,
+// optionally overridden by the given overrides (typically from the URL of the
+// /oauth2/start request) according to the configured rules for this provider.
+func (p *ProviderData) LoginURLParams(overrides url.Values) url.Values {
+	// the returned url.Values may be modified later in the request handling process
+	// so shallow clone the default map
+	params := url.Values{}
+	for k, v := range p.loginURLParameterDefaults {
+		params[k] = v
+	}
+	if len(overrides) > 0 {
```

Comment:
Would it make sense instead to loop over this list in the other way, ie if the for loop was `for param, values := range overrides`? Just wondering if that's likely to be more efficient, or, if we could then un-nest some of this at all if we were to do that

Reply:
`p.loginURLParameterOverrides` is the same size for every request, typically no more than two or three elements but in many (most?) cases it will actually be an empty map.  Whereas `overrides` is `req.URL.Query()` so it will be different for each request and could in the worst case be arbitrarily large (at least up to the limit imposed by the max length of a URL query string).  So I'd say in _most_ cases it'll be more efficient the way it is now, and even in the worst case the hit will be consistently small for every request.

Reply:
That seems like a good argument for keeping it this way, SGTM!